migration/v2.25/README.md

```diff
@@ -0,0 +1,97 @@
+- [ ] Check if there are any pending changes to the environment;
+- [ ] Check the state of the environment, pods, nodes and backup jobs:
+
+    ```bash
+    ./compliantkubernetes-apps/bin/ck8s test sc|wc
+    ./compliantkubernetes-apps/bin/ck8s ops kubectl sc|wc get pods -A -o custom-columns=NAMESPACE:metadata.namespace,POD:metadata.name,READY-false:status.containerStatuses[*].ready,REASON:status.containerStatuses[*].state.terminated.reason | grep false | grep -v Completed
+    ./compliantkubernetes-apps/bin/ck8s ops kubectl sc|wc get nodes
+    ./compliantkubernetes-apps/bin/ck8s ops kubectl sc|wc get jobs -A
+    velero get backup
+    ```
+
+- [ ] Silence the notifications for the alerts. e.g you can use [alertmanager silences](https://prometheus.io/docs/alerting/latest/alertmanager/#silences);
+
+## Steps that can be done before the upgrade - non-disruptive
+
+1. Checkout the new release: `git switch -d v2.25.x-ck8sx`
+
+1. Switch to the correct remote: `git submodule sync`
+
+1. Update the kubespray submodule: `git submodule update --init --recursive`
+
+1. Run `bin/ck8s-kubespray upgrade both v2.25 prepare` to update your config.
+
+    > [!NOTE]
+    > It is possible to update `wc` and `sc` config separately by replacing `both` when running the `upgrade` command, e.g. the following will only update config for the workload cluster:
+    >
+    > ```bash
+    > bin/ck8s-kubespray upgrade wc v2.25 prepare
+    > ```
+
+1. Download the required files on the nodes
+
+    ```bash
+    ./bin/ck8s-kubespray run-playbook sc upgrade_cluster.yml -b --tags=download
+    ./bin/ck8s-kubespray run-playbook wc upgrade_cluster.yml -b --tags=download
+    ```
+
+1. Run `migration/v2.25/prepare/20-change-topology-constraints.sh` **or** manually add the following snippet at the end of both `sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` and `wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml`
+
+    ```yaml
+    kube_scheduler_profiles:
+      - schedulerName: default-scheduler
+        pluginConfig:
+          - name: PodTopologySpread
+            args:
+              defaultingType: List
+              defaultConstraints:
+                - maxSkew: 1
+                  topologyKey: kubernetes.io/hostname
+                  whenUnsatisfiable: ScheduleAnyway
+                - maxSkew: 1
+                  topologyKey: topology.kubernetes.io/zone
+                  whenUnsatisfiable: ScheduleAnyway
+
+    ```
+
+## Upgrade steps
+
+These steps will cause disruptions in the environment.
+
+1. Upgrade the cluster to a new kubernetes version:
+
+    ```bash
+    ./bin/ck8s-kubespray run-playbook sc upgrade_cluster.yml -b -e skip_downloads=true
+    ./bin/ck8s-kubespray run-playbook wc upgrade_cluster.yml -b -e skip_downloads=true
+    ```
+
+1. Restart `kube-scheduler` on the control plane nodes:
+
+    ```bash
+    sudo bash -c "crictl pods  --name 'kube-scheduler*' -q | xargs -I% bash -c 'crictl stopp % && crictl rmp %'"
+    ```
```

Comment:
Question: is it not enough to restart the "kube-scheduler-<control-plane>" K8s pods?

Reply:
Good question.

As I understood from @Xartos and some further experimentation confirms, `kube-scheduler` is not a real pod and attempting to delete it with `kubectl` shows (after turning up the log level) the scheduler receiving this event and then recreating the virtual pod. It doesn't appear to look at config files.

The [specific method is copied from an event handler in kubespray](https://github.com/elastisys/kubespray/blob/1a73b254578b5e5f9765510a5d40d0038376b54d/roles/kubernetes/control-plane/handlers/main.yml#L35-L44) that unfortunately is not triggered by the config file having changed.

Reply:
Yea, it's a static pod so it can't be restarted with `kubectl`
However you should be able to use ansible to make this migration a bit smoother

```suggestion
1. Restart `kube-scheduler` on the control plane nodes:

    ```bash
    TERRAFORM_STATE_ROOT="${CK8S_CONFIG_PATH}/sc-config"
ansible -i "${CK8S_CONFIG_PATH}/sc-config/inventory.ini" kube_control_plane -b -m shell -a "crictl pods  --name 'kube-scheduler*' -q | xargs -I% bash -c 'crictl stopp % && crictl rmp %'"
  TERRAFORM_STATE_ROOT="${CK8S_CONFIG_PATH}/wc-config"
ansible -i "${CK8S_CONFIG_PATH}/wc-config/inventory.ini" kube_control_plane -b -m shell -a "crictl pods  --name 'kube-scheduler*' -q | xargs -I% bash -c 'crictl stopp % && crictl rmp %'"
    ```
```